From 066fee2e7d6834f24838bc1896aa38ca77209e3c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Caol=C3=A1n=20McNamara?= <caolanm@redhat.com>
Date: Thu, 16 Mar 2017 15:53:53 +0000
Subject: [PATCH] honor max_memory_to_use if its set

I'd like to use JPEGMEM to limit memory that libjpeg will allocation
to reject sizes that would the default 2G limit under asan while
fuzzing LibreOffice's jpeg integration
---
 jmemnobs.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/jmemnobs.c b/jmemnobs.c
index 5797198..c7dc560 100644
--- a/jmemnobs.c
+++ b/jmemnobs.c
@@ -66,14 +66,24 @@ jpeg_free_large (j_common_ptr cinfo, void *object, size_t sizeofobject)
 
 /*
  * This routine computes the total memory space available for allocation.
- * Here we always say, "we got all you want bud!"
  */
 
 GLOBAL(size_t)
 jpeg_mem_available (j_common_ptr cinfo, size_t min_bytes_needed,
                     size_t max_bytes_needed, size_t already_allocated)
 {
-  return max_bytes_needed;
+  if (!cinfo->mem->max_memory_to_use)
+  {
+    /* Here we always say, "we got all you want bud!" */
+    return max_bytes_needed;
+  }
+
+  if (cinfo->mem->max_memory_to_use - already_allocated >= max_bytes_needed)
+  {
+      return max_bytes_needed;
+  }
+
+  return cinfo->mem->max_memory_to_use - already_allocated;
 }
 
 
-- 
2.9.3

